### PR TITLE
Removes some unnecessary adds

### PR DESCRIPTION
**Pull request by jon-chuang**
## Description

Perf impact is minimal but still worth doing for completeness.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: ec/src/models/short_weierstrass_jacobian.rs [hunk 1]
```diff
@@ -106,7 +106,12 @@ impl<P: Parameters> GroupAffine<P> {
     #[allow(dead_code)]
     pub fn get_point_from_x(x: P::BaseField, greatest: bool) -> Option<Self> {
         // Compute x^3 + ax + b
-        let x3b = P::add_b(&((x.square() * &x) + &P::mul_by_a(&x)));
+        // Rust does not optimise away addition with zero
+        let x3b = if P::COEFF_A == P::BaseField::zero() {
```
Comment (Pratyush) [inline ec/src/models/short_weierstrass_jacobian.rs:110]:
```suggestion
        let x3b = if P::COEFF_A.is_zero() {
```

File: ec/src/models/short_weierstrass_jacobian.rs [hunk 2]
```diff
@@ -122,7 +127,12 @@ impl<P: Parameters> GroupAffine<P> {
         } else {
             // Check that the point is on the curve
             let y2 = self.y.square();
-            let x3b = P::add_b(&((self.x.square() * &self.x) + &P::mul_by_a(&self.x)));
+            // Rust does not optimise away addition with zero
+            let x3b = if P::COEFF_A == P::BaseField::zero() {
```
Comment (Pratyush) [inline ec/src/models/short_weierstrass_jacobian.rs:131]:
```suggestion
            let x3b = if P::COEFF_A.is_zero() {
```

File: ec/src/models/mod.rs [hunk 3]
```diff
@@ -29,9 +29,12 @@ pub trait SWModelParameters: ModelParameters {
 
     #[inline(always)]
     fn add_b(elem: &Self::BaseField) -> Self::BaseField {
-        let mut copy = *elem;
-        copy += &Self::COEFF_B;
-        copy
+        if Self::COEFF_B != Self::BaseField::zero() {
```
Comment (Pratyush) [inline ec/src/models/mod.rs:32]:
```suggestion
        if !Self::COEFF_B.is_zero() {
```

**Comment by Pratyush** [comment]
Is this useful elsewhere too? Should we incorporate this directly in `Field::add/mul/sub?`

File: ec/src/models/short_weierstrass_jacobian.rs (hunk 1 above)
Comment (ValarDragon) [reply ec/src/models/short_weierstrass_jacobian.rs:110]: Is it valid syntax to do something like
```
let mut x3b: P::Basefield;
if (P::COEFF_A.is_zero()) {
    x3b = P::add_b(&(x.square() * &x));
} else {
    x3b = P::add_b(&((x.square() * &x) + &P::mul_by_a(&x)));
}
```
imo assigning to an if is always confusing syntax.

**Comment by ValarDragon** [comment]
LGTM pending the change to is_zero

**Review by huitseeker — APPROVED** [review]

**Review by weikengchen — APPROVED** [review]
Merged with prior comments.

**Pull request merged**